subsys/bluetooth/audio/pacs.c

```diff
@@ -54,124 +81,87 @@ static void pac_data_add(struct net_buf_simple *buf, uint8_t num,
 	}
 }
 
-static void get_pac_records(struct bt_conn *conn, enum bt_audio_dir dir,
-			    struct net_buf_simple *buf)
+struct pac_records_build_data {
+	struct net_buf_simple *buf;
+	uint8_t num_pac;
+};
+
+static bool build_pac_records(const struct bt_audio_capability *capability, void *user_data)
 {
-	struct bt_pacs_read_rsp *rsp;
+	struct pac_records_build_data *data = user_data;
+	struct net_buf_simple *buf = data->buf;
+	struct bt_pac_meta *meta;
+	struct bt_codec *codec;
+	struct bt_pac *pac;
 
-	/* Reset if buffer before using */
-	net_buf_simple_reset(buf);
+	codec = capability->codec;
 
-	rsp = net_buf_simple_add(buf, sizeof(*rsp));
-	rsp->num_pac = 0;
+	pac = net_buf_simple_add(buf, sizeof(*pac));
```

Comment:
We are adding quite a lot of data to the buffer here, but we have no checks for size. This seems dangerous as we do not restrict the number of capabilities we can register. 

If we register 50 capabilities, but can only fit 25 in the buffer (which actually should be `MIN(CONFIG_BT_L2CAP_TX_MTU, BT_ATT_MAX_ATTRIBUTE_LEN)`), should we just put the first 25 in the response, or do something else? 

Alternatively we need to verify the total length of the PAC records when registering capabilities, so that we do no register more than what we can respond with. 

This can be postponed to a different PR, as this is just a move of existing code :)